From d188e722d444a7cced0d51f2593be857d1c4ceab Mon Sep 17 00:00:00 2001
From: Rudi <r.ihle@s-t.de>
Date: Sun, 3 Aug 2014 10:41:47 +0200
Subject: [PATCH 1014/1020] ASoC: fsl_hdmi/imx-hdmi-dma: Enable/fix 24bit
 audio output.

Since SNDRV_PCM_FORMAT_S24 uses 4 bytes, the constraint has to be set to 32.
Also we need to extend the parity calculation to full 32 bit.

Signed-off-by: Rudi <r.ihle@s-t.de>
---
 sound/soc/fsl/fsl_hdmi.c     |    2 +-
 sound/soc/fsl/imx-hdmi-dma.c |    1 +
 2 files changed, 2 insertions(+), 1 deletion(-)

diff --git a/sound/soc/fsl/fsl_hdmi.c b/sound/soc/fsl/fsl_hdmi.c
index f190e8c..92b9c7d 100644
--- a/sound/soc/fsl/fsl_hdmi.c
+++ b/sound/soc/fsl/fsl_hdmi.c
@@ -301,7 +301,7 @@ static void fsl_hdmi_get_playback_sample_size(void)
 	playback_sample_size[i++] = 16;
 
 	if (edid_cfg.sample_sizes & 0x4)
-		playback_sample_size[i++] = 24;
+		playback_sample_size[i++] = 32;
 
 	playback_constraint_bits.list = playback_sample_size;
 	playback_constraint_bits.count = i;
diff --git a/sound/soc/fsl/imx-hdmi-dma.c b/sound/soc/fsl/imx-hdmi-dma.c
index 059df83..2ddc3c4 100644
--- a/sound/soc/fsl/imx-hdmi-dma.c
+++ b/sound/soc/fsl/imx-hdmi-dma.c
@@ -213,6 +213,7 @@ static void hdmi_mask(int mask)
 
 static inline int odd_ones(unsigned a)
 {
+	a ^= a >> 16;
 	a ^= a >> 8;
 	a ^= a >> 4;
 	a ^= a >> 2;
-- 
1.7.9.5

